### ISSUE TITLE
u2net跟unet訓練到一定次數Precision跟Recall都會變0

### ISSUE THREAD

**Issue by SuPoTing** [PaddlePaddle/PaddleSeg#2998]
### 问题确认 Search before asking

- [X] 我已经查询[历史issue](https://github.com/PaddlePaddle/PaddleSeg/issues)(包括open与closed)，没有发现相似的bug。I have searched the [open and closed issues](https://github.com/PaddlePaddle/PaddleSeg/issues) and found no similar bug report.

### Bug描述 Describe the Bug

2023-02-23 22:39:49 [INFO]      [TRAIN] epoch: 15, iter: 37000/50000, loss: 1.6889, lr: 0.002975, batch_cost: 0.2060, reader_cost: 0.00000, ips: 4.8547 samples/sec | ETA 00:44:37
2023-02-23 22:39:49 [INFO]      Start evaluating (total_samples: 1264, total_iters: 1264)...
1264/1264 [==============================] - 95s 75ms/step - batch_cost: 0.0749 - reader cost: 1.0532e-04
2023-02-23 22:41:24 [INFO]      [EVAL] #Images: 1264 mIoU: 0.4906 Acc: 0.9813 Kappa: 0.0000 Dice: 0.4953
2023-02-23 22:41:24 [INFO]      [EVAL] Class IoU:
[0.9813 0.    ]
2023-02-23 22:41:24 [INFO]      [EVAL] Class Precision:
[0.9813 0.    ]
2023-02-23 22:41:24 [INFO]      [EVAL] Class Recall:
[1. 0.]
2023-02-23 22:41:24 [INFO]      [EVAL] Class DICE:
[0.9906 0.    ]

### 复现环境 Environment

Windows10 專業版
GPU 3090
paddlepaddle 2.4.2
paddlepaddle-gpu 2.4.1psot11.7
paddleseg2.7
Python3.8.16
Cuda11.7
cudnn8.4.1

### Bug描述确认 Bug description confirmation

- [ ] 我确认已经提供了Bug复现步骤、代码改动说明、以及环境信息，确认问题是可以复现的。I confirm that the bug replication steps, code change instructions, and environment information have been provided, and the problem can be reproduced.

### 是否愿意提交PR？ Are you willing to submit a PR?

- [X] 我愿意提交PR！I'd like to help by submitting a PR!

**Comment by zhaoyantao-murray** [comment]
您需要提供下使用的yml配置文件和训练命令等信息

**Comment by SuPoTing** [comment]
u2net_cityscapes_1024x512_160k.yml
====================================
_base_: '../_base_/custom.yml'

batch_size: 1
iters: 10000
MIRROR: True

model:
  type: U2Net
  num_classes: 2
  pretrained: Null

loss:
  coef: [1, 1, 1, 1, 1, 1, 1]

custom.yml
====================================
batch_size: 1
iters: 10000
save_interval: 500
train_dataset:
  type: Dataset
  dataset_root: C:/Users/SUPOTING/Desktop/PaddleSeg2.7/dataset
  train_path: C:/Users/SUPOTING/Desktop/PaddleSeg2.7/dataset/train.txt
  num_classes: 2
  transforms:
    - type: ResizeStepScaling
      min_scale_factor: 0.5
      max_scale_factor: 2.0
      scale_step_size: 0.25
    - type: RandomPaddingCrop
      crop_size: [1024, 512]
    - type: RandomHorizontalFlip
    - type: RandomDistort
      brightness_range: 0.4
      contrast_range: 0.4
      saturation_range: 0.4
    - type: Normalize
  mode: train

val_dataset:
  type: Dataset
  dataset_root: C:/Users/SUPOTING/Desktop/PaddleSeg2.7/dataset
  val_path: C:/Users/SUPOTING/Desktop/PaddleSeg2.7/dataset/val.txt
  num_classes: 2
  transforms:
    - type: Resize
      target_size: [1024, 512]
    - type: Normalize
  mode: val


optimizer:
  type: sgd
  momentum: 0.9
  weight_decay: 4.0e-5

lr_scheduler:
  type: PolynomialDecay
  learning_rate: 0.01
  end_lr: 0
  power: 0.9

loss:
  types:
    - type: DiceLoss
  coef: [1]

指令
====================================
python train.py --config configs/u2net/u2net_cityscapes_1024x512_160k.yml --do_eval --save_dir output/u2net_original/ --iters 50000 --num_workers 12

**Comment by SuPoTing** [comment]
主要是做兩類而已我用segformer訓練不會有這種情況

**Comment by zhaoyantao-murray** [comment]
> 主要是做兩類而已我用segformer訓練不會有這種情況

想问下你的两类是指前景和背景这两类吗？还是类别1，类别2，背景类的意思呢？

**Comment by SuPoTing** [comment]
前景跟背景而已

**Comment by zhaoyantao-murray** [comment]
> 前景跟背景而已

那有可能是U2net和Unet这两个模型相对较大，参数量多，在训练自己的数据集的时候过拟合，模型预测会都偏向预测背景类，所以前景类的指标就变成零了。segformer是轻量级的模型，没有过拟合，可以正常分割出前景类。

**Comment by SuPoTing** [comment]
我之前有使用2080ti訓練沒這個問題
所以我懷疑是不是Paddle沒有對3090優化呢?

**Comment by zhaoyantao-murray** [comment]
不太可能是硬件的原因，建议先排查一下是否是环境问题，先跑一下quick start 
https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.7/docs/quick_start_cn.md
是否能正常收敛。

**Comment by SuPoTing** [comment]
> 不太可能是硬件的原因，建议先排查一下是否是环境问题，先跑一下quick start https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.7/docs/quick_start_cn.md 是否能正常收敛。

我發現DiceLoss有問題 使用CrossEntropyLoss沒問題

**Comment by zhaoyantao-murray** [comment]
> 

你的意思是PaddleSeg的Diceloss实现有问题呢，还是说只是使用Diceloss会导致类别1的eval指标为0呢？
如果是后者的话，应该是DiceLoss对于你的任务是不合适的，使用Cross-EntropyLoss是更合适的；
如果是前者的话可以麻烦你指出是具体哪里代码实现有问题嘛？

**Comment by SuPoTing** [comment]
是後者，但我也訓練類似黃斑部病變的資料
用quick start測試DiceLoss也有這個問題

**Comment by lichengkong** [comment]
> 是後者，但我也訓練類似黃斑部病變的資料 用quick start測試DiceLoss也有這個問題

u2net 最后使用7个Loss相加，你更换loss为dice 之后只保留了一个loss，而实际上loss个数只与模型的输出结构有关。应当是7个。按理说这里应当是断言报错无法训练才对。

loss:
  types:
    - type: CrossEntropyLoss
  coef: [1, 1, 1, 1, 1, 1, 1]

loss:
types:
  - type: DiceLoss
coef: [1]

**Comment by marshall-dteach** [comment]
&emsp;&emsp;检查一下你的`label`，我怀疑你的`label`是背景像素是0和前景像素是255，建议用`img=cv2.imread(file)`读取文件，使用`print(np.unique(img))`来验证一下，如果结果是`[0, 255]`而非`[0, 1]`，在读取`label`后统一除255。

**Comment by shiyutang** [comment]
这个问题我更倾向觉得可能在于DICELoss和CELoss的特性不同导致，因此判断具体是哪种情况的话，可以进一步在正确医学图像数据集是否也会出现类似的现象
